Fix debug logs failing with persistent connection

ganeshrn opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Fixes #33047

~*  As debug logs are written on stdout, it interrupts
   the communication between ansible-connection(background)
  process and main process. To avoid this add a string delimiter to exactly 
  identify the response string.~

*  ansible-connection which runs as a background process sends a
   json string (contains response received from remote device)
   to foreground ansible-playbook process over stdout.

*  If in case debug flag is enabled the connection_loader api
   invoked from ansible-connection `ssh = connection_loader.get('ssh', class_only=True)`
   results in emitting debug logs on stdout. This  spurious log
   interfere with the actual response and results in failure while
   reading json string in ansible-playbook process

* To avoid this save stdout of ansible-connection and redirect it string
  buffer to accumulate all the logs emitted by core API's

* Add these logs in `result['messages']` which is send a json string after reinstating saved stdout

*  Remove unwanted code in ansible-connection

* Add `stdin.flush()` after write to  handle buffer scenario in case of PY3
 
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->

 - Bugfix Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
bin/ansible-connection
executor/task_executor.py
plugins/connection/persistent.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

ganeshrn committed:
Fix debug logs failing with persistent connection

Fixes #33047

*  As debug logs are written on stdout, it interrupts
   the communication between ansible-connection(background)
   process and main process. To avoid this add a string similar
   to exactly identify the response string.

*  Remove unwanted code in ansible-connection

*  Fix review comments

Qalthos commented:
This does work around the problem, but loses all the debug info. I think I'd rather seen ansible-connection take over stdout/stderr so those lines can get pushed back as messages properly, rather than just dropping them.

ganeshrn commented:
I didn't get your comment. AFAIK ansible-connection sends the request to cliconf/netconf plugin over jsonrpc and waits for the response which is in json format. However, while it is waiting a spurious debug log is send over stdout which is also considered as part of the response.
Within `ansible-connection` probably there is no way to differentiate between debug log and json response.

ganeshrn commented:
If debug logs can be extrated in `executor/task_executor.py` and copy it to messages.
```
logs = out.split(br"##RESPONSE##:")[0]
out = out.split(br"##RESPONSE##:")[1]
```

kedarkekan commented on lib/ansible/executor/task_executor.py:
```diff
@@ -817,10 +817,12 @@ def _start_connection(self):
+            return None
+
+        out = out.split(br"##RESPONSE##:")[1]
+        result = json.loads(to_text(out.rstrip(), errors='surrogate_then_replace'))
```
please use out.strip() to handle leading whitespaces from messages too.

ganeshrn replied:
Fixed

kedarkekan commented on lib/ansible/plugins/connection/persistent.py:
```diff
@@ -92,10 +93,12 @@ def _start_connection(self):
+
+        out = out.split(br"##RESPONSE##:")[1]
+        result = json.loads(to_text(out.rstrip(), errors='surrogate_then_replace'))
 
```
Let us print the verbose logs as well? 
`for line in output[0].splitlines():`

ganeshrn replied:
Fixed

bdowling commented:
Is this ready to be merged?  It addressed the issue for me.  Thanks!

ganeshrn commented:
@bdowling The approach in the fix is not correct as message passed between background and foreground process is expected to be a json string. Adding a delimiter before json string defeats the purpose.

I am working on an alternate approach to fix this issue in a better way.

ganeshrn committed:
Fix spurious log emitted on ansible-connection stdout issue

*  ansible-connection which runs as a background process sends a
   json string (contains response received from remote device)
   to foreground ansible-playbook process over stdout.

*  If in case debug flag is enabled the connection_loader api
   invoked from ansible-connection `ssh = connection_loader.get('ssh', class_only=True)`
   results in emitting debug logs on stdout. This  spurious log
   interfere with the actual response and results in failure while
   reading json string in ansible-playbook process

* To avoid this save stdout of ansible-connection and redirect it string
  buffer to accumulate all the logs emitted by core API's

* Add these logs in `result['messages']` which is send a json string after reinstating saved stdout

*  Remove unwanted code in ansible-connection

ganeshrn committed:
Fix review comment

ganeshrn commented:
ready_for_review
